### ISSUE TITLE
[bug]: Endpoints periodically deemed Unreachable after responding to OPTIONS request

### ISSUE THREAD

**Issue by zylinc-ms** [asterisk/asterisk#835]
### Severity

Major

### Versions

20.6.0

### Components/Modules

pjsip_transport_events

### Operating Environment

debian:bookworm

### Frequency of Occurrence

Frequent

### Issue Description

I have a customer that has two siptrunks connections to two SBC's - both using TLS.
Our address is configured in the SBC's so we do not register towards these SipTrunks.
We are using Asterisk version 20.6.0.

Frequently, but not always, Asterisk marks the endpoints as unreachable immediately after we resceive their response to an OPTIONS request sent by us.
When the agents then tries to answer or transfer any calls, our requests are rejected with the "Is endpoint registered and reachable?" reason.

The qualify_frequency is set to 60, and when we 1 minute later send a new OPTIONS request, Asterisk now believes that the endpoint is reachable again, and the agents can now work again.
I see no pattern for when this happens, but typically happens 5-6 times per day and other days not at all. 
It happens both when there is high load on the system, and early moring when the (almost) only activity is OPTIONS requests and responses.

I have attached the full log and the pjsip configuration for the TLS Transport and the SipTrunks.
The full log is collected with debug level 3 and verbose at level 5.
Immediately before the endpoints are deemed unrereachable, I see these log entries from pjsip_transport_events.c:
 - RefCnt: 5 state:DISCONNECTED
 - RefCnt: 5 state:SHUTDOWN
 - RefCnt: 5 : state:MONITOR_DESTROYED
These are not present when the responses for our OPTIONS requests are accepted.

I would guess that they indicate that the TCP connection is dropped, and there is a race condition that leads to the described error.
I do not have any TCP dump proving that though, so I hope that the full log has the necessary infomation.

There are 2 incidents in the attached log where one of the customer’s siptrunks were deemed as unreachable:

[Jul 11 13:44:31] SBC1 - Unreachable
[Jul 11 13:45:31] SBC1 - Reachable
[Jul 11 14:03:31] SBC1 - Unreachable
[Jul 11 14:04:31] SBC1 - Reachable

I haven't seen this problem on any other of our customers that are set up in the same way.
I fear that the reason is just that this customer handles a massive amount of calls compared to our other customers, that might not notice that the system is broken.

Please let me know if you need any additional information.

[full.txt](https://github.com/user-attachments/files/16548080/full.txt)
[Pjsip_transport.txt](https://github.com/user-attachments/files/16548081/Pjsip_transport.txt)
[Pjsip_siptrunks.txt](https://github.com/user-attachments/files/16548083/Pjsip_siptrunks.txt)

### Asterisk Issue Guidelines

- [X] Yes, I have read the Asterisk Issue Guidelines

**Comment by gtjoseph** [comment]
The DISCONNECTED state actually comes from pjproject when it detects an issue when trying to connect or send or receive data. It look like something happened to the sbc1 connection after we received the 200 OK response from the OPTIONS which caused the connection to drop.  That in turn caused the endpoint to go unavailable.  A few seconds later sbc1 sent asterisk an OPTIONS request to which asterisk responded with a 200 OK but that doesn't trigger asterisk to do anything else.  So, it's not until a minute later when the qualify_frequency timer expires that asterisk attempts to send another OPTIONS that the connection gets re-established and the endpoint becomes available again.  Unfortunately, there's no way to tell from the logs exactly why the connection had issues in the first place.

A few things...

* A PCAP would really help determine exactly what's happening at the network level.
* The timestamp resolution in the log file is only to the "seconds" level so that makes it difficult to see when things are really happening.  Setting `dateformat=%F %T.%3q %z` in the `[general]` section of logger.conf would be a help.
* The output of `pjsip show settings` would help, particularly the value of `keep_alive_interval` in the `[global]` section of pjsip.conf.  The default is 90 seconds which is not very useful if your qualify frequency is already 60 seconds.  You could try setting `keep_alive_interval` to 10 seconds or even 5 seconds.  It's low impact because it's just a packet with `\r\n\r\n` in it.  This won't re-establish the connection but it may help keep the connection up.
* Starting with Asterisk versions 20.8.0 and 21.3.0, you can now set tcp-layer keepalives for individual pjsip transports.  These are even less impact than the `\r\n\r\n` ones and may also help keep the connection up.

Having said all of that, there are some things that we could be doing better in this situation...
* If we do see a connection drop, we could try re-qualifying immediately to try and re-establish the connection without waiting for the next scheduled qualify.  Maybe with some additional options to say... If the connection drops, wait X seconds, then try requalifying Y times at Z second intervals.
* If the remote end connects back to Asterisk, also immediately restart the qualification process on the outbound side if Asterisk's connection to the remote had been previously disconnected.
* We may be able to have the keepalive process try to re-establish the connection if it had dropped.

We'll be looking into these but they wouldn't be quick fixes.

The bottom line though is still understanding why the connection is dropping in the first place.  Whatever it is, it doesn't seem to be affecting sbc2 at all.

**Comment by zylinc-ms** [comment]
Hi gtjoseph

Thank you for your response.

Please note that it is not just the SBC1 that has the problem. It happens just as often for SBC2. Typically twice a day for each SBC, but some days not at all.

Please also note that since we are using TLS, the OPTIONS requests are not sent from the port that we accept SIP requests on. As you can see from the rport parameter in the responses to our options request, the portnumber was 58723 in the first example and 55825 in the second.
I have now set the keepalive interval to 10, but won't the keepalive messages also be sent from a dynamic port, just like all other SIP requests?

I don't know whether it is Asterisk or the remote server that tears down the socket connection, but there is certainly no reason to conclude that the remote server is unreachable. We are still listening on 5067 as written in our Contact header and as configured on both SBCs. Asterisk will not reuse the connection it made when sending the OPTIONS request, so it should be safe for either party to close it as soon as the response has been sent.

I have re-enabled the Qualify for both SIP trunks, started a tcpdump, and will let you know as soon as I get a new incident.

[PjSipSettings.txt](https://github.com/user-attachments/files/16674645/PjSipSettings.txt)

**Comment by zylinc-ms** [comment]
I caught 2 new incidents - one for each SBC:

2024-08-21 07:09:45.207 SBC1 Unreachable
2024-08-21 07:10:45.214 SBC1 Reachable
2024-08-21 07:58:39.667 SBC2 Unreachable
2024-08-21 07:59:39.707 SBC2 Reachable

Full log and PCAP attached .
The PCAP is filtered on the sip port of the 2 SBCs

[Full_Unreachable.zip](https://github.com/user-attachments/files/16689725/Full_Unreachable.zip)
[Unreachable.zip](https://github.com/user-attachments/files/16689726/Unreachable.zip)

**Comment by gtjoseph** [comment]
I set up a lab environment to test but can't seem to find a way to reproduce this.  I think a good PCAP is really what we need.

**Comment by gtjoseph** [comment]
I have to ask this...  Do you have any patches applied to asterisk or do you use any modules not in the asterisk source tree?  I ask because as I'm trying to recreate this issue, if I set up a TLS connection between two asterisk instances just `kill -KILL` one of them, the one that's still up immediately shows the same DISCONNECTED, SHUTDOWN, etc. messages from pjsip_transport_events but then within less than a second it retries.  If the other asterisk has already come back up, the new transport gets established and the endpoint state never goes to unreachable.

The same thing happens if I use the `tcpkill` program to just kill the tcp connection without disturbing the asterisk instances themselves.  

This is on 20.6.0 with the system, global, aor, endpoint and identify settings you provided.

**Comment by zylinc-ms** [comment]
I'm sorry about the wrong period for the PCAP I sent you.
I forgot that wireshark shows me the timestamps in my local time, and not UTC as the full log shows.
I have attached new full log with matching PCAP, and you will now see a TCP Reset when the connection is broken followed by a new SYN ACK.
First unreachable incident is around frame 6616, and the second one at frame 10151.

The remaining times the connection is broken, without the OPTIONS becoming a problem can be seen in the following frames: 
4276, 5812, 6450, 7201, 9616, 9851, 11916 and 12549.

I haven't changed the outbound proxy, because they will be rewritten the next time somebody make changes to the configuration, and I doubt that this is part of the problem. To be honest I can't fully remember the reason for the outbound proxy, other than that I had an issue for TCP and TLS connections when I didn't include the proxy with the transport=tls/tcp ending.

And yes, I have made some changes to the source code. Mainly in app_queue.c where I have added additional features to offer the caller the possibility to be called back when it is their turn, and introduced another position to be announced, since we often prioritize the calls differently than pure FIFO.
Additionally I have changed say.c to not concatenate sound files when saying ex. the number 79. The concatenation works ok for English but sounds terrible in Danish, so we have all the numbers from 1-99 as individual sound files for all our supported languages.
I have never introduced any new modules, or changed anything within PjSIP though.

One thing that is different for this tenant is that Asterisk almost never is idle, but is handling many calls throughout the day. The number of simultaneous calls is not that high and is very rarely higher than 10, but there are only very few periods where it is not handling calls at all, and I have never seen an unreachable incident during idle time.
Additionally whenever someone makes configuration changes in our portal, the config files are rewritten followed by a core reload command. 
I am just mentioning this because we previously experienced crashes when the transports were configured with "allow_reload=true". The crash did not happen upon reload, but happened when one of the calls that were active during the reload was disconnected.

I really appreciate your efforts to mimic our setup, but maybe the trick is to have another Asterisk sending calls to the system, and periodically performing a reload.
I hope that the new PCAP will shed some light on the problem.

[UnreachableLog.zip](https://github.com/user-attachments/files/16708732/UnreachableLog.zip)
[UnreachableCapture.zip](https://github.com/user-attachments/files/16708734/UnreachableCapture.zip)

**Comment by gtjoseph** [comment]
Well, this is interesting.  Just before those connections reset, asterisk has sent a TLS Encrypted Alert with cause 21 "Decryption of a TLSCiphertext record is decrypted in an invalid way: either it was not an even multiple of the block length or its padding values, when checked, were not correct. This message is always fatal."  This would indicate that asterisk got a corrupted packet.  Most of the time, the transport and monitor just shut down without causing the endpoint to go unavailable and the next outgoing request re-establishes the connection.  However, it seems that if the reset happens at some very specific time, the reset ripples up to the endpoint and causes it to go unavailable.  This gives me something to go on.   However...

If you look at the log you'll notice that right around the same time as the resets occur for the SBC transports, there are SSL errors for websocket clients.  Run `grep -E "(SSL error|MONITOR_DESTROYED)" UnreachableLog.txt` and you'll see them all.  I think you have a larger network issue going on that's manifesting itself as data corruption.  You don't mention much about the environment other than it's "debian bookworm".  Is this a virtualized environment or bare metal?

**Comment by gtjoseph** [comment]
OK, I can actually reproduce the issue now.  If a corrupted packet is received and the reset occurs while the qualify is in progress, the qualify will fail and that's why the endpoint goes unreachable.  Even if we've actually received and started processing the 200 OK, if the transport disconnects before we've finished processing it, it'll interrupt the process and cause the qualify to fail.  This is going to take some thought to fix so it'snot going to be quick.  Your best bet at this point is to determine why you're getting the corrupted packets in the first place.

**Comment by zylinc-ms** [comment]
Yes I noticed the MONITOR_DESTROYED events, which are always present when the connection is deemed unreachable.
By far most of them are caused by the agents using WebRTC as their device, and they can connect and disconnect seemingly at random. I assume that it was someone clicking F5 or just shutting down the browser without existing nicely. I of course can't categorically rule out that this is pointing at a larger problem, but I assume the problem most likely are crappy connections or crappy agents. When they use WebRTC we contact them directly and do not route the call via the SBCs.
All the WebRTC agents have a name starting with "web_" - like PJSIP/web_nicolas.rath_bestseller.com.
I guess you can guess the "algorithm" for the remaining part of the name :-)

The environment is highly virtualized. We are running them in Docker Swarm mode within Azure. 
Each customer has their own container with Asterisk (and a lot of other modules each of them also in their own container). 
All containers except for Asterisk are communicating solely via RabbitMQ.  
We seperate the Asterisk containers on ports - so Tenant1 uses ports 5062 for udp/tcp and 5063 for tls. Tenant2 uses ports 5064 and 5065 and so on. Per default we reserve 400 ports for each customer for the RTP.
We rarely have more than 40 customers per swarm, so running out of ports is actually not a problem.
I at first was sceptical about this approach, but it is performing amazingly well, 
So it is not a classical multitenant setup, but rather multiinstance, which makes deployment of new updates a lot easier, beacuse we can gurantee that it happens at a specific time that suits the individual customers.

**Comment by zylinc-ms** [comment]
Great to hear that you can reproduce it :-)
That certainly makes it easier.

As stated above, I am not convinced that we have a major problem with corrupted packages, but rather have a large portion of agents behind their private wi-fi at home - which we naturally can't control.
I am quite sure that the SBCs are top of the line and sufficiently specced to handle the communication, and I have many times loadtested Asterisk by sending a new call each second, each call lasting for 3 minutes. The number of simultanous calls are always kept at a steady 180 with absolutely no audio problems.

**Comment by gtjoseph** [comment]
Well, the connection resets for the SBCs are definitely being caused by the deliberate closing of the connection due to the TLS Encrypted Alerts which are always fatal.  Do you have access to the SBC logs to see what they report when the resets happen?

It might be interesting for you to get a pcap that includes the webrtc clients and see if their disconnects are being caused by TLS Alerts or just because the TCP connection drops.

**Comment by zylinc-ms** [comment]
I will have access to SBC logs when it happens again, and I have included the WebRTC sip port in the trace that is running now.
But "unfortunately" there has been no Unreachable events yet.
I have only enabled Qualify for SBC2 so the customer will be less affected. They use SBC1 when answering calls.

Regardless of what the logs show, do you see a way to handle this racecondition?
If you already are in the process of reading a response to an OPTIONS request, then you know that something was returned, and the SipTrunk is not unreachable - regardless of the event from PjSip.

I don't know that part of the code, and I am sure that it is not as simple as it sounds.
I am only asking because I fear that we have the same problem for many tenants, and the only reason we haven't heard about is that the customer isn't as busy as this one, or that they just conclude that our product isn't that good.

**Comment by gtjoseph** [comment]
We do have an internal issue open with a "major" severity but I'm not sure when it can be worked.  I still encourage you to look at your network to understand why the TLS alerts are happening in the first place.  How many hops are there between the asterisk instances and the SBC? If you use `ping` or `mtr` with various `-i`, `-c` and `-s` options, what's the RTT and packet loss? Can you stress test the path between them?, etc.

**Comment by zylinc-ms** [comment]
Hi George
I am sorry that I haven't replied yet.
Unfortunately the customer is currently considering whether they want to replace our solution with one of our competitors.
I expect that we will be able to keep most of the countries where they have offices, but at least for now I will unfortunately not be able to provide traces from the SBCs.

Since we are deploying our solution with Docker, where many containers exists on the same machines, I haven't been able to get something usefull out of mtr og ping.

The internal "major" issue you mentioned is that this one, or do you have bigger plans where this could be part of it?
And will you let me know here, or is there a certain release version that I should keep my eye on?

For now I have disabled the OPTIONS keepalives, and hoping the keep_alive_interval can help keeping the line open.

**Comment by jcolp** [comment]
The internal issue is this one, and there is no specific release it is targeted for. It is just in queue along with everything else to be looked at.

**Comment by ksjoberg** [comment]
I have seen this surface on 20.17 through 20.19 as well as 22.9.0, and I've now traced the exact mechanism.

### Root cause: DNS queue contention between concurrent INVITE and OPTIONS transactions

Asterisk resolves the PJSIP remote_hosts hostname through `dns_system_resolver_tp`, a single-threaded task processor. Each SIP transaction that needs to resolve a hostname submits a set of parallel DNS queries (`NAPTR`, three `SRV` variants, and an `A` record fallback) to that thread as independent tasks. When an outbound `INVITE` happens to be dispatched at roughly the same time as the qualify timer fires, two full sets of DNS queries (ten tasks in total) are queued on that single thread. The `INVITE`'s resolver drains its queries first, and the OPTIONS resolver's queries run only after, incurring the full serialization delay on top of its own resolution time.

In the specific failure I captured, the `INVITE`'s `DNS` resolver was created 677 ms before the `OPTIONS` resolver and completed normally in 4.21 seconds. The `OPTIONS` resolver, processing the exact same hostname, took 6.43 seconds (roughly the INVITE's time plus the queue backlog) and fired its completion callback 355 ms after the 6-second tsx timer had already expired and marked the endpoint `Unreachable`. When the `OPTIONS` packet was finally transmitted it received an immediate 200 OK, confirming the remote endpoint itself is fully reachable; it is purely a timeout race.

This is consistent with what I see across all qualify cycles in the logs: when no concurrent `INVITE` is in flight, DNS for this hostname resolves in ~4.2 seconds and the qualify succeeds with roughly 1.8 seconds to spare. That margin disappears entirely when call setup happens to coincide with a qualify.

### The actionable points:

* `qualify_timeout` covers the entire qualify round-trip including DNS, not just the `OPTIONS` exchange itself. In environments that resolve a hostname (rather than using a static IP), the DNS time must be budgeted into this setting. A value significantly above the worst-case DNS time plus one RTT is necessary for reliable qualification.
* Using a static IP or a `/etc/hosts` entry for the PJSIP peer `remote_host` in environments where hostname resolution latency is significant (e.g. Docker container-to-container resolution) eliminates the DNS component entirely and makes the qualification deterministic.
* Alternatively, caching/pre-resolving DNS for the PJSIP contact before the qualify fires . This way, the `OPTIONS` transaction starts with a warm cache hit rather than issuing fresh queries, and would remove the dependency on `dns_system_resolver_tp` queue depth.

**Comment by gtjoseph** [comment]
Using a local caching resolver is also an option.  named, unbound, systemd-resolved, and others.

**Comment by ksjoberg** [comment]
I completely expected a local caching resolver would not be necessary as the name that takes time to resolve is the name of a companion docker container running on the same host. It should hit docker's resolver at 127.0.0.11 but this time is apparently not negligible.
